api/envoy/api/v2/cds.proto

```diff
@@ -150,12 +151,24 @@ message Cluster {
   // when picking a host in the cluster.
   LbPolicy lb_policy = 6 [(validate.rules).enum.defined_only = true];
 
-  // If the service discovery type is
+  // Setting this as a requirement for clusters with discovery type
   // :ref:`STATIC<envoy_api_enum_value_Cluster.DiscoveryType.STATIC>`,
   // :ref:`STRICT_DNS<envoy_api_enum_value_Cluster.DiscoveryType.STRICT_DNS>`
-  // or :ref:`LOGICAL_DNS<envoy_api_enum_value_Cluster.DiscoveryType.LOGICAL_DNS>`,
-  // then hosts is required.
-  repeated core.Address hosts = 7;
+  // or :ref:`LOGICAL_DNS<envoy_api_enum_value_Cluster.DiscoveryType.LOGICAL_DNS>`
+  // is deprecated. Set the :ref:`endpoints<envoy_api_field_Cluster.endpoints>` field
+  // instead.
+  repeated core.Address hosts = 7 [deprecated = true];
+
+  // Setting this is required for specifying members of
+  // :ref:`STATIC<envoy_api_enum_value_Cluster.DiscoveryType.STATIC>`,
+  // :ref:`STRICT_DNS<envoy_api_enum_value_Cluster.DiscoveryType.STRICT_DNS>`
+  // or :ref:`LOGICAL_DNS<envoy_api_enum_value_Cluster.DiscoveryType.LOGICAL_DNS>` clusters.
+  // This field supersedes :ref:`hosts<envoy_api_field_Cluster.hosts>` field.
+  //
+  // .. attention::
+  //
+  //   Setting this overrides :ref:`hosts<envoy_api_field_Cluster.hosts>` values.
```

Comment:
I have a higher level API issue I'd like to raise if we're going to add this. From time-to-time, folks wonder if we can provide the full set of EDS capabilities embedded in a CDS response (i.e. in a `Cluster`). To do this, we would effectively need to have a `LocalityLbEndpoints` here, rather than `repeated Endpoint`. It would close one weird inconsistency in the Envoy API, where you can basically get all the behaviors with a static bootstrap config _except_ for some EDS special ones (e.g. locality balancing). @mattklein123 for thoughts as well.

Reply:
@htuch @mattklein123 could we keep both of them ((`core.Address`) `hosts` and (`endpoint.LocalityLbEndpoints`) `endpoints`)? 

Reason is, I'm not sure if it makes sense: by having `hosts` will be easier to digest for e.g. a newcomer like me. It's simple and the depth of the config is shallow for specifying cluster members "statically".

A config accommodating `endpoint.LocalityLbEndpoints` will be something like:

```yaml
    endpoints:
      lb_endpoints:
      - endpoint:
          address:
            socket_address:
              address: localhost1
              port_value: 11001
          health_check_config:
            port_value: 8000
```

Which is not only deep but it also has that `lb_endpoints`. IMO, not super obvious at the first glance for defining cluster members statically.

Reply:
Agree with @htuch that this is a good opportunity to fix this inconsistency. In general, I agree with @dio that the lb_endpoints construct is more confusing, however, we also optimize for machine generated config. My feeling is to support lb_endpoints, and just have a good example in the docs of a sample configuration. I think with that it should be pretty straightforward for people to use.

Reply:
@htuch/@mattklein123 should it be `repeated LocalityLbEndpoints` as in eds?

Reply:
Yeah, repeated as in https://github.com/envoyproxy/envoy/blob/872728d4ad50acececaeae389da59cb72846cf89/api/envoy/api/v2/eds.proto#L41. Maybe it should just be the `ClusterLoadAssignment`, since that has additional cross-locality policy embedded in it as well.

Reply:
OK. So we will treat the current `Cluster.name` to be the same as `ClusterLoadAssignment.cluster_name`?

Reply:
@htuch can we consider this, in EDS context, as a config update (`onConfigUpdate`)?

Reply:
@dio not sure if I fully grok the question. Are you asking if we can use `onConfigUpdate` to process the `ClusterLoadAssignment` when embedded in a `Cluster`? If so, I don't think you use exactly that method, but you can factor out the logic to somewhere shared between EDS and `Cluster` load, sure, it's the same code in both cases.

Reply:
@htuch yes, I meant to take out the logic from eds and use that for other clusters. 

I have another question, while refactoring the `StaticClusterImpl`, I failed to satisfy these conditions: https://github.com/envoyproxy/envoy/blob/master/test/common/upstream/upstream_impl_test.cc#L646-L648. Could you enlight me about those assertions a bit more (why they have to be zero)? Thanks!

Reply:
An interesting aspect of this PR is that we now support logical DNS in embedded ClusterLoadAssignments. This didn't use to work, we had a number of questions about this IIRC.